### ISSUE TITLE
gtk::timeout_add and friends should probaby take a Duration instead

### ISSUE THREAD

**Issue by upsuper** [gtk-rs/gtk#990]
It seems to me that it makes more sense to have `gtk::timeout_add` and `glib::timeout_add{,_local}` take a [`Duration`](https://doc.rust-lang.org/std/time/struct.Duration.html) and remove their `_seconds` counterparts.

This would make it clearer on the meaning of the parameter, and matches better to what the standard library does, thus feels more idiomatic.

**Comment by GuillaumeGomez** [comment]
I don't see how it'd be more idiomatic to expect a type that we don't full use. It's even worse than that: we'd take a type and don't make it being used *correctly* since we don't use all information.

**Comment by upsuper** [comment]
What do you mean by "don't use all information"? I think these functions pretty much means to take a duration, so the typing is more correct.

If you mean that `Duration` takes nanosecond, but you would be dropping such precision, I think you can just document to make it clear what precision is expected. [`sleep`](https://doc.rust-lang.org/std/thread/fn.sleep.html) function in the standard library takes `Duration`, while apparently not all systems support nanosecond-level sleeping, and thus the precision is dropped on some systems like [Windows](https://github.com/rust-lang/rust/blob/45d050cde277b22a755847338f2acc2c7b834141/src/libstd/sys/windows/mod.rs#L286-L300).

**Comment by sdroege** [comment]
I agree that using `Duration` for such things would be the right thing to do. My main worry is that we should do it consistently, and that might require some changes in the code generator so it's also done for automatically generated code.

To what other functions would this apply? There are certainly more functions that have some kind of timeout, time duration parameter.

**Comment by upsuper** [comment]
As normally a duration parameter is just an integer, and you would need to be able to parse its semantics, I'm not sure it's really possible to automate. Maybe it's more practical to manually exclude them from auto-generated code?

And yeah, we would need to make it consistent, so it needs some extra investigation to identify other functions doing so. But this doesn't feel like something common in the code, so I wouldn't be surprised if there aren't many :)

**Comment by sdroege** [comment]
I would suggest to add something to the Gir.toml for this to configure that some function parameter (or return value!) is actually a time duration, and then it uses `Duration` plus conversion. We already have a few such things.

**Comment by GuillaumeGomez** [comment]
I misunderstood your initial comment, my bad... Then I guess a `Duration` would be fine.
